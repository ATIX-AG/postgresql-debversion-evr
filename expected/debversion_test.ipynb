{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This file generates outputfile for debversioncompare tests using the built in version compare from dpkg."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subprocess import run\n",
    "\n",
    "def compare(version1, version2, op):\n",
    "    dictionary = {\"<\":\"lt\", \">\":\"gt\", \"<=\":\"le\", \">=\":\"ge\", \"=\":\"eq\", \"!=\":\"ne\"}\n",
    "    dpkg_op = dictionary[op] \n",
    "    command = f\"dpkg --compare-versions {version1} {dpkg_op} {version2}\"\n",
    "    result = run(command.split())\n",
    "    if result.returncode == 0:\n",
    "        return 't'\n",
    "    else:\n",
    "        return 'f'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'t'"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare('1.1.1', '1.1.2', '<')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "with open(\"../sql/debversion_test.sql\") as f:\n",
    "    fin = f.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "output = open(\"debversion_test.out\", \"wt\")\n",
    "for line in fin:\n",
    "    if line == '\\n':\n",
    "        continue\n",
    "    output.write(line)\n",
    "    if line.startswith(\"SELECT\"):\n",
    "        mask = re.compile(r\"SELECT '([0-9a-zA-Z\\+\\-\\.\\:\\~]+)'::debversion ([\\>\\<\\=\\!]+) '([0-9a-zA-Z\\+\\-\\.\\:\\~]+)'::debversion;\")\n",
    "        match = mask.match(line)\n",
    "        version1 = match.group(1)\n",
    "        op = match.group(2)\n",
    "        version2 = match.group(3)\n",
    "        letter = compare(version1, version2, op)\n",
    "        text = f\" ?column? \\n----------\\n {letter}\\n(1 row)\\n\\n\"\n",
    "        output.write(text)\n",
    "\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comparing function using debian python implementation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "from debian import debian_support \n",
    "def compare1(version1, version2, op):   \n",
    "    vc = debian_support.version_compare(a,b)\n",
    "    if vc > 0:\n",
    "        print('version a > version b')\n",
    "        if op in [\">\",\">=\"]:\n",
    "            return \"t\"\n",
    "        else:\n",
    "            return \"f\"\n",
    "    elif vc == 0:\n",
    "        print('version a == version b')\n",
    "        if op  == \"=\":\n",
    "            return \"t\"\n",
    "        else:\n",
    "            return \"f\"\n",
    "    elif vc < 0:\n",
    "        print('version a < version b')\n",
    "        if op in [\"<\", \"<=\"]:\n",
    "            return \"t\"\n",
    "        else:\n",
    "            return \"f\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
