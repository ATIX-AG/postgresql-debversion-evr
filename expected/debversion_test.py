#!/usr/bin/env python3
# coding: utf-8

# # This file generates outputfile for debversioncompare tests using the built in version compare from dpkg.

from subprocess import run
import re
import apt_pkg

FILES=[ 'debversion_test', 'test_case_from_dpkg', 'brute-force-test' ]

apt_pkg.init_system()

def compare_lib(version1, version2):
    res = apt_pkg.version_compare(version1, version2)

    if (res > 0):
        return 1;
    elif (res < 0):
        return -1;
    else:
        return 0;

def compare_cmd(version1, version2, op):
    dictionary = {"<":"lt", ">":"gt", "<=":"le", ">=":"ge", "=":"eq", "!=":"ne"}
    dpkg_op = dictionary[op]
    command = f"dpkg --compare-versions {version1} {dpkg_op} {version2}"
    result = run(command.split())
    if result.returncode == 0:
        return 't'
    else:
        return 'f'

def create_output(line):
    mask = re.compile(r"SELECT deb_version_cmp\('([0-9a-zA-Z\+\-\.\:\~]+)', '([0-9a-zA-Z\+\-\.\:\~]+)'\);")
    match = mask.match(line)
    version1 = match.group(1)
    version2 = match.group(2)
    letter = compare(version1, version2)
    return f" deb_version_cmp \n-----------------\n              {str(letter).rjust(2)}\n(1 row)\n\n"

# CHANGEME to use apt-pkg library or dpkg command for comparing
compare = compare_lib

for filename in FILES:
    print(f"Processing {filename}...", end='')
    try:
        with open(f"../sql/{filename}.sql") as fin:
            output = open(f"{filename}.out", "wt")
            for line in fin:
                if line == '\n':
                    continue
                output.write(line)
                if line.startswith("SELECT deb_version_cmp('1:bla-1'"):
                  output.write("WARNING:  version 1:bla-1 has bad syntax: version number does not start with digit\n")
                if line.startswith("SELECT"):
                    output.write(create_output(line))
            output.close()
        print('done')
    except IOError:
        print("File not accessible")
