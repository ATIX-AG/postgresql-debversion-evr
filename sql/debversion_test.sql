CREATE EXTENSION debversion_evr;

--equal versions
SELECT deb_version_cmp('1:2.3-bla-1', '1:2.3-bla-1');

-- different epoch
SELECT deb_version_cmp('1:2.3-bla-1', '2.3-bla-1');

-- different revision
SELECT deb_version_cmp('1:2.3-bla-1', '1:2.3-bla-2');

--different versions
SELECT deb_version_cmp('1:2.3-bla-1', '1:2.3-blaa-1');

--real world examples
SELECT deb_version_cmp('1.60-26+b1', '1.60+git20161116.90da8a0-1');

--non-valid versions (but dpkg can handle these with warning message)
SELECT deb_version_cmp('1:bla-1', '1:2.3-bla-1');

SELECT deb_version_cmp('1:2.3-bla-a', '1:2.3-bla-1');

--these throw errors in dpkg (dpkg cannot handle them)
--SELECT deb_version_cmp('1:2.3-bla-1', '1:2.3-bla-');

--long numeric values in version
SELECT deb_version_cmp('1.0.20181217162649+dfsg-10', '1.0.20181217162649+dfsg-10');
