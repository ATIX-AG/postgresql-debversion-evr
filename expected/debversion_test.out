CREATE EXTENSION debversion_evr;
--equal versions
SELECT deb_version_cmp('1:2.3-bla-1', '1:2.3-bla-1');
 deb_version_cmp 
-----------------
               0
(1 row)

-- different epoch
SELECT deb_version_cmp('1:2.3-bla-1', '2.3-bla-1');
 deb_version_cmp 
-----------------
               1
(1 row)

-- different revision
SELECT deb_version_cmp('1:2.3-bla-1', '1:2.3-bla-2');
 deb_version_cmp 
-----------------
              -1
(1 row)

--different versions
SELECT deb_version_cmp('1:2.3-bla-1', '1:2.3-blaa-1');
 deb_version_cmp 
-----------------
              -1
(1 row)

--real world examples
SELECT deb_version_cmp('1.60-26+b1', '1.60+git20161116.90da8a0-1');
 deb_version_cmp 
-----------------
              -1
(1 row)

--non-valid versions (but dpkg can handle these with warning message)
SELECT deb_version_cmp('1:bla-1', '1:2.3-bla-1');
WARNING:  version 1:bla-1 has bad syntax: version number does not start with digit
 deb_version_cmp 
-----------------
               1
(1 row)

SELECT deb_version_cmp('1:2.3-bla-a', '1:2.3-bla-1');
 deb_version_cmp 
-----------------
               1
(1 row)

--these throw errors in dpkg (dpkg cannot handle them)
--SELECT deb_version_cmp('1:2.3-bla-1', '1:2.3-bla-');
--long numeric values in version
SELECT deb_version_cmp('1.0.20181217162649+dfsg-10', '1.0.20181217162649+dfsg-10');
 deb_version_cmp 
-----------------
               0
(1 row)

